FirebaseDatabase/Sources/Realtime/FWebSocketConnection.m

```diff
@@ -255,26 +346,52 @@ - (void)handleIncomingFrame:(NSString *)message {
 }
 
 #pragma mark -
+#pragma mark URLSessionWebSocketDelegate watchOS implementation
+#if TARGET_OS_WATCH
+
+- (void)URLSession:(NSURLSession *)session
+          webSocketTask:(NSURLSessionWebSocketTask *)webSocketTask
+    didOpenWithProtocol:(NSString *)protocol {
+    [self webSocketOpened];
+}
+
+- (void)URLSession:(NSURLSession *)session
+       webSocketTask:(NSURLSessionWebSocketTask *)webSocketTask
+    didCloseWithCode:(NSURLSessionWebSocketCloseCode)closeCode
+              reason:(NSData *)reason {
+    FFLog(@"I-RDB083011", @"(wsc:%@) didCloseWithCode: %ld %@",
+          self.connectionId, (long)closeCode, reason);
+    [self onClosed];
+}
+
+- (void)receiveWebSocketData {
+    [self.webSocketTask receiveMessageWithCompletionHandler:^(
+                            NSURLSessionWebSocketMessage *_Nullable message,
+                            NSError *_Nullable error) {
+      if (message) {
+          [self handleIncomingFrame:message.string];
+      } else if (error && !isClosed) {
+          FFWarn(@"I-RDB083020",
+                 @"Error received from web socket, closing the connection. %@",
+                 error);
+          [self shutdown];
+          return;
+      }
+
+      [self receiveWebSocketData];
+    }];
+}
+
+#else
+
 #pragma mark SRWebSocketDelegate implementation
+
 - (void)webSocket:(FSRWebSocket *)webSocket didReceiveMessage:(id)message {
     [self handleIncomingFrame:message];
 }
 
 - (void)webSocketDidOpen:(FSRWebSocket *)webSocket {
-    FFLog(@"I-RDB083008", @"(wsc:%@) webSocketDidOpen", self.connectionId);
-
-    everConnected = YES;
-
-    dispatch_async(dispatch_get_main_queue(), ^{
-      self->keepAlive =
-          [NSTimer scheduledTimerWithTimeInterval:kWebsocketKeepaliveInterval
-                                           target:self
-                                         selector:@selector(nop:)
-                                         userInfo:nil
-                                          repeats:YES];
-      FFLog(@"I-RDB083009", @"(wsc:%@) nop timer kicked off",
-            self.connectionId);
-    });
+    [self webSocketOpened];
```

Comment:
We own FSRWebSocketDelegate and could just remove the argument here, which would remove the need for this indirection.

Reply:
Fixed in latest commit.